### PR TITLE
30 add default location when creating a server

### LINKED ISSUE #30: Add default location when creating a server
When a server gets created a default location gets added with the path "/". 

If a location is specified in the config file with the path "/" the default location values get changed accordingly to the ones provided .

### LINKED ISSUE #92: Default values: ConfigServer
Set following values as default in the constructor of ConfigServer:

- Host: 0.0.0.0
- Port: 8080

### LINKED ISSUE #93: Translate localhost to 127.0.0.1
If "localhost" is specified as the host in a configuration file, it should be stored as '127.0.0.1'.

### PR DESCRIPTION
**Pull request by QCHR1581**
# TL;DR

- Translating `localhost` to `127.0.0.1`
- Parsing the path of a location block
- Adding default `Location` when `ConfigServer` object gets created
- Adding values of location block with the path `/` to the default location
- Inheritance: server &rarr; location

## Translating localhost to 127.0.0.1

When `localhost` gets specified as host in the config file, the ip `127.0.0.1` is stored as host. 
The function `isIpAddressValid` was updated to accept `localhost`.

##  Parsing the path of a location block

The path of a location block is getting parsed and stored in the corresponding `Location` object. 

## Adding default `Location` when `ConfigServer` object gets created

When a `ConfigServer` object gets created,  a default `Location` is added with the path `/`. 

## Adding values of location block with the path `/` to the default location

If a location with the path `/` gets specified in the config file, the values of this location are stored in the default `Location`. 
Therefore **no** new `Location` is added in this case to the vector of `Location`. 

## Inheritance: server &rarr; location

A server block and a location block share following directives:

- root
- client_max_body_size
- error_page

If no values are specified for these directives in the location block, they inherit their values from the server block.

## Tests

For all implementations test were created. 

Concerning the topic *Inheritance: server &rarr; location*:

Currently only the inheritance for the root directive gets tested. The reason for this is that the other two directives get not parsed yet. 

--- 
Closes #93 
Closes #92
Closes #30

### CODE REVIEW

File: src/ConfigFileParser.cpp [hunk 1]
```diff
@@ -389,28 +440,33 @@ void ConfigFileParser::readSocket(const std::string& value)
...
-
+			std::string ipAddress = value.substr(0, semicolonIndex);
+			if (ipAddress == "localhost")
+				m_configFile.servers[m_serverIndex].host = "127.0.0.1";
+			else {
+				std::string port = value.substr(0, semicolonIndex);
+				if (!webutils::isPortValid(port))
+					throw std::runtime_error("Invalid port");
+				m_configFile.servers[m_serverIndex].port = port;
+			}
```
Comment (u413-284-si) [inline src/ConfigFileParser.cpp:463]: Here you are assuming if there is no '.' found in the string that it must be either
1. localhost
2. port
And you are extracting the same substring twice and assigning it to two different strings. I would just do it once in the beginning and give it a more general name, e.g. temp or endpoint.

Optionally could add a comment as well to explain this as it is not evident on first sight.

File: test/test_ConfigFileParser.cpp [hunk 2]
```diff
@@ -366,10 +383,26 @@ TEST_F(ValidConfigFileTests, DirectiveAndClosingBracketOnSameLine)
...
+	EXPECT_EQ("/hello", configFile.servers[0].locations[1].path);
+	EXPECT_EQ("/world", configFile.servers[0].locations[2].path);
+}
+
+TEST_F(ValidConfigFileTests, DirectiveInheritance)
+{
+	ConfigFile configFile;
+	EXPECT_NO_THROW(configFile = m_configFileParser.parseConfigFile("config_files/directive_inheritance.conf"));
+	EXPECT_EQ("/var/www/html", configFile.servers[0].locations[0].root);
+}
```
Comment (u413-284-si) [inline test/test_ConfigFileParser.cpp:408]: The PR is missing both config files. Therefore the tests are segfaulting:

image

**Review by u413-284-si — CHANGES_REQUESTED** [review]
Good work! Some improvements can be made and a couple files are missing.

File: src/ConfigFileParser.cpp [hunk 3]
```diff
@@ -330,6 +347,38 @@ bool ConfigFileParser::readAndTrimLine(const std::string& content, char delimite
...
+{
+	std::string directiveLocation = "location";
+	size_t startIndex = directiveLocation.length();
+
+	while (std::isspace(m_currentLine[startIndex]) != 0)
+		startIndex++;
+
+	size_t endIndex = startIndex;
+	while (std::isspace(m_currentLine[endIndex]) == 0)
+		endIndex++;
```
[the remark this answers is not in the document]
Comment (QCHR1581) [reply src/ConfigFileParser.cpp]: Good idea, but I only want to remove the leading spaces from the string. 

For example:

`location /data {`

I save the length of `location` in a var so I can skip this part. 
After this step, I want to remove/skip the whitespaces to reach the `/` and then save the remaining non whitespaces chars. Which means saving `/data`. 

I rewrote the code now to also use the `trimLeadingWhitespaces` function

File: src/ConfigFileParser.cpp (hunk 1 above)
Comment (QCHR1581) [reply src/ConfigFileParser.cpp]: True, could be confusing. 

I use now only the var `hostOrPort` for storing the substring.

File: src/ConfigFileParser.cpp [hunk 4]
```diff
@@ -330,6 +347,37 @@ bool ConfigFileParser::readAndTrimLine(const std::string& content, char delimite
...
+ * If the path is "/", the m_locationIndex will be set to 0 because the default location has an index of 0
+ * Therefore the following functions will store the values correctly in the default location
+ *
+ * Otherwise a new location will be created, added to the locations vector and the m_locationIndex will be incremented
+ *
+ */
+void ConfigFileParser::readLocationBlockPath(void)
+{
+	std::string directiveLocation = "location";
+	size_t startIndex = directiveLocation.length();
```
Comment (gwolf-011235) [inline src/ConfigFileParser.cpp:362]: This will always be 8. Just having 8 would be hard to read.
You could use sizeof() to not create intermediate string and eval at compile time: 
`const size_t startIndex = sizeof("location") - 1; //subtract zero terminator`
Or just a constant: `const size_t startIndex = 8; // Length of "location"`

**Review by gwolf-011235 — APPROVED** [review]
Looks good to me!

Just commented on a few little things (like missing newline), which you can resolve on your own :D

**Review by u413-284-si — APPROVED** [review]

File: config_files/listen_localhost_and_port.conf [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    server {
+        listen localhost:9090;
+        server_name example.com;
+        root /var/www/html;
+
+        location / {
+            autoindex on;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (QCHR1581) [reply config_files/listen_localhost_and_port.conf]: Done

File: config_files/listen_only_localhost.conf [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    server {
+        listen localhost;
+        server_name example.com;
+        root /var/www/html;
+
+        location / {
+            autoindex on;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (QCHR1581) [reply config_files/listen_only_localhost.conf]: Done

File: config_files/location_path.conf [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+		location /world {
+            autoindex on;
+        }
+
+		location / {
+			autoindex on;
+		}
+    }
+}
```
[the remark this answers is not in the document]
Comment (QCHR1581) [reply config_files/location_path.conf]: Done

File: src/ConfigFileParser.cpp [hunk 8]
```diff
@@ -279,20 +275,41 @@ void ConfigFileParser::processServerContent(const ServerBlockConfig& serverBlock
  * The content will be processed by reading it line by line (delimited by ';')
  * In this process the values of the directives will be read and stored
  *
+ * If the path of the location block is "/", the m_locationIndex will be set to 0 in readLocationBlockPath
+ * This is necessary because the following functions need to store the parsed values in the the default location then
```
[the remark this answers is not in the document]
Comment (QCHR1581) [reply src/ConfigFileParser.cpp]: True, is done

File: src/ConfigFileParser.cpp (hunk 4 above)
Comment (QCHR1581) [reply src/ConfigFileParser.cpp]: Good point. I also did not want to write just 8.

I think your first suggestion fits in this case good and I used it for the purpose of calculating the index value.
